File: assets/code-lang/bst.java
public class bst { //Define a public class named bst
    private Node root; //Declare a private instance variable "root" of type node

    private class Node { //Define a private inner class named "Node"
        Node left; //Declare a left child Node
        Node right; //Declare a right child Node
        int data; //Declare a data variable of type int

        public Node(int newData) { //Constructor for the Node class, initializes left, right, and data
            left = null;
            right = null;
            data = newData;
        }
    }

    public void binarySearchTree() { //constructor for the bst class, initializes the root to null
        root = null; //creates empty binary tree
    }

    public void add(int data) { //public method to add a node with the given data to the bst
        root = add(root, data); //call the private add method with the root and the given data
    }
    private Node add(Node node, int data) { //private method to recursively add a node with the given data to the bst
        if(node == null) { //checks if the current node is null
            node = new Node(data); //create a new node with the data if so
        }
        else {
            if(data <= node.data) { //if the data is less than or equal to the current node's data
                node.left = add(node.left, data); //recursively add to the left subtree
            }
            else { //if the data is greater
                node.right = add(node.right, data); //recursively add to the right subtree
            }
        }
        return node; //return the modified code
    }

    public void remove(int data) { //public method to remove a node with the given data to the bst
        root = remove(root, data); //call the private remove method with the root and the given data
    }

    private Node remove(Node node, int data) { // Private method to remove a node with the given data from the binary search tree
        if(node == null) { // Base case: if the current node is null, return null (no change)
            return null;
        }
        if(data < node.data) {  // If the data to be deleted is smaller than the root's data, then it lies in the left subtree
            node.left = remove(node.left, data);  // Recursively call remove on the left subtree
        }
        else if(data > node.data) {  // If the data to be deleted is greater than the root's data, then it lies in the right subtree
            node.right = remove(node.right, data);  // Recursively call remove on the right subtree
        }
        // If the data to be deleted is the same as the root's data,
        // then this is the node to be deleted
        else {
            if(node.left == null && node.right == null) { // Case 1: Node with no children
                return null; // Simply remove the node by returning null
            }
            if(node.left == null) { // Case 2: Node with only a right child
                return node.right; // Replace the node with its right child
            }
            else if(node.right == null) { // Case 3: Node with only a left child
                return node.left; // Replace the node with its left child
            }
            else {  // Case 4: Node with two children
                Node maxNode = findMaxValueNode(node.left);  // Find the maximum value node in the left subtree
                node.data = maxNode.data;  // Replace the current node's data with the maxNode's data
                node.left = remove(node.left, maxNode.data);  // Recursively remove the maxNode from the left subtree
            }

        }
        return node;
    }
    private Node findMaxValueNode(Node node) { // Private method to find and return the node with the maximum value in the binary search tree
        if(node == null) { // Base case: if the current node is null, the tree is empty, return null
            return null;
        }
        while(node.right != null) { // Traverse to the right child until the right child is null, which is the maximum value node
            node = node.right;
        }
        return node; // Return the node with the maximum value
    }
     private void inOrderTraversal(Node node) {
        if(node != null) {
            inOrderTraversal(node.left);
            System.out.print(node.data + " ");
            inOrderTraversal(node.right);
        }
    }
    public static void main(String[] args) {
        // Create an instance of the bst class
        bst binarySearchTree = new bst();

        // Test the add method
        binarySearchTree.add(50);
        binarySearchTree.add(30);
        binarySearchTree.add(70);

        // Print the tree to verify the add operation
        System.out.println("Binary Search Tree after adding elements:");
        binarySearchTree.inOrderTraversal(binarySearchTree.root);

        // Test the remove method
        binarySearchTree.remove(30);

        // Print the tree to verify the remove operation
        System.out.println("\nBinary Search Tree after removing 30:");
        binarySearchTree.inOrderTraversal(binarySearchTree.root);

    }
}

File: assets/code-lang/bst.py
# Define a class representing a Node in the binary search tree
class Node:
    # Constructor to initialize a Node with data, left, and right attributes
    def __init__(self, data):
        self.data = data  # Initialize the data attribute with the given data value
        self.left = None  # Initialize the left child as None
        self.right = None  # Initialize the right child as None


# Define a class representing a Binary Search Tree (BST)
class BST:
    # Constructor to initialize the BST with a root attribute
    def __init__(self):
        self.root = None  # Initialize the root as None

    # Public method to add a node with the given data to the BST
    def add(self, data):
        self.root = self._add(self.root, data)  # Call the private _add method with the root and the given data

    def _add(self, node, data):  # Private method to recursively add a node with the given data to the BST
        if node is None:  # Check if the current node is None
            return Node(data)  # Create a new node with the given data if the current node is None
        else:
            # If the data is less than or equal to the current node's data, recursively add to the left subtree
            if data <= node.data:
                node.left = self._add(node.left, data)
            else:
                node.right = self._add(node.right, data)  # If the data is greater, recursively add to the right subtree
        return node

    def remove(self, data):  # Public method to remove a node with the given data from the BST
        self.root = self._remove(self.root, data)  # Call the private _remove method with the root and the given data

    def _remove(self, node, data):  # Private method to recursively remove a node with the given data from the BST
        if node is None:  # Check if the current node is None
            return None  # Base case: if the current node is None, return None (no change)
        # If the data to be deleted is smaller than the root's data, recursively call _remove on the left subtree
        elif data < node.data:
            node.left = self._remove(node.left, data)
        # If the data to be deleted is greater than the root's data, recursively call _remove on the right subtree
        elif data > node.data:
            node.right = self._remove(node.right, data)
        else:
            # If the data to be deleted is the same as the root's data, handle different cases
            if node.left is None and node.right is None:
                return None  # Case 1: Node with no children, simply remove the node by returning None
            elif node.left is None:
                return node.right  # Case 2: Node with only a right child, replace the node with its right child
            elif node.right is None:
                return node.left  # Case 3: Node with only a left child, replace the node with its left child
            else:
                # Case 4: Node with two children, find the maximum value node in the left subtree
                max_node = self._find_max_value_node(node.left)
                node.data = max_node.data  # Replace the current node's data with the max_node's data
                # Recursively remove the max_node from the left subtree
                node.left = self._remove(node.left, max_node.data)
        return node

    @staticmethod
    def _find_max_value_node(node):  # Private method to find and return the node with the maximum value in the BST
        if node is None:  # Base case: if the current node is None, the tree is empty, return None
            return None
        # Traverse to the right child until the right child is None, which is the maximum value node
        while node.right is not None:
            node = node.right
        return node

    def _in_order_traversal(self, node):  # Private method for in-order traversal of the BST and print the nodes
        if node is not None:
            self._in_order_traversal(node.left)  # Recursively traverse the left subtree
            print(node.data, end=" ")  # Print the current node's data
            self._in_order_traversal(node.right)  # Recursively traverse the right subtree

    def in_order_traversal(self, root):
        pass


# Main block executed when the script is run
if __name__ == "__main__":
    # Create an instance of the BST class
    binary_search_tree = BST()

    # Test the add method
    binary_search_tree.add(50)
    binary_search_tree.add(30)
    binary_search_tree.add(70)

    # Print the tree to verify the add operation
    print("Binary Search Tree after adding elements:")
    binary_search_tree._in_order_traversal(binary_search_tree.root)  # Use _in_order_traversal for this verification

    # Test the remove method
    binary_search_tree.remove(30)

    # Print the tree to verify the remove operation
    print("\nBinary Search Tree after removing 30:")
    binary_search_tree._in_order_traversal(binary_search_tree.root)  # Use _in_order_traversal for this verification